src/test/java/bio/terra/service/snapshotbuilder/query/QueryTest.java

```diff
@@ -122,4 +127,52 @@ void renderComplexSQL() {
                 "(c.condition_concept_id = 316139 OR c0.ancestor_concept_id = 316139 OR c.condition_concept_id = 4311280 OR c0.ancestor_concept_id = 4311280)"),
             containsString("AND p.year_of_birth < 1983")));
   }
+
+  @Test
+  void renderSQLWithDatasetModel() {
+    DatasetModel dataset =
+        new DatasetModel().name("name").dataProject("project").cloudPlatform(CloudPlatform.GCP);
+    TablePointer conceptTablePointer =
+        new TablePointer("concept", null, null, generateTableName(dataset));
+    TableVariable conceptTableVariable = TableVariable.forPrimary(conceptTablePointer);
+    FieldPointer nameFieldPointer = new FieldPointer(conceptTablePointer, "concept_name");
+    FieldVariable nameFieldVariable = new FieldVariable(nameFieldPointer, conceptTableVariable);
+    FieldPointer idFieldPointer = new FieldPointer(conceptTablePointer, "concept_id");
+    FieldVariable idFieldVariable = new FieldVariable(idFieldPointer, conceptTableVariable);
+
+    TablePointer tablePointer =
+        new TablePointer("concept_ancestor", null, null, generateTableName(dataset));
+    TableVariable tableVariable = TableVariable.forPrimary(tablePointer);
+    FieldPointer fieldPointer = new FieldPointer(tablePointer, "descendant_concept_id");
+    FieldVariable fieldVariable = new FieldVariable(fieldPointer, tableVariable);
+    BinaryFilterVariable whereClause =
+        new BinaryFilterVariable(
+            new FieldVariable(new FieldPointer(tablePointer, "ancestor_concept_id"), tableVariable),
+            BinaryFilterVariable.BinaryOperator.EQUALS,
+            new Literal(100));
+    Query subQuery = new Query(List.of(fieldVariable), List.of(tableVariable), whereClause);
+    SubQueryFilterVariable subQueryFilterVariable =
+        new SubQueryFilterVariable(idFieldVariable, SubQueryFilterVariable.Operator.IN, subQuery);
+    Query query =
+        new Query(
+            List.of(nameFieldVariable, idFieldVariable),
+            List.of(conceptTableVariable),
+            subQueryFilterVariable);
+    String sql = query.renderSQL();
+    assertThat(
+        sql,
+        is(
+            "SELECT c.concept_name, c.concept_id FROM `project.datarepo_name.concept` AS c "
+                + "WHERE c.concept_id IN "
+                + "(SELECT c.descendant_concept_id FROM `project.datarepo_name.concept_ancestor` AS c "
+                + "WHERE c.ancestor_concept_id = 100)"));
+  }
+
+  public static Function<String, String> generateTableName(DatasetModel dataset) {
```

Comment:
Where should this method really live once it is actually being used in queries?

Reply:
I'm not sure, but it should be shared with the code in `BigQueryVisitor`. One way would be to make a static method in `BigQueryVisitor` that does this, that accepts `dataset` and `tableName`, which then both this code and `BigQueryVisitor` could use.